### ISSUE TITLE
LSP: add support for server requests

### ISSUE THREAD

**Issue by tboby** [georgewfraser/fsharp-language-server#7]
The LSP has a server to client request `workspace/applyEdit`. I also use custom requests to do things that are easier using vscode's typescript api than through the LSP.

In https://github.com/tboby/cwtools-vscode/tree/master/src/LSP I adapted it to add a queue for handling client responses. It's quite hacky though.

I can try to put it together into a pull request if you'd like, but I suspect you might have a preferred alternative approach.

**Comment by georgewfraser** [comment]
I actually noticed that you did that, and your code was quite helpful when I implemented server => client messages. I used a [similar approach](https://github.com/georgewfraser/fsharp-language-server/blob/master/src/LSP/LanguageServer.fs#L183), though I think a few details were different. I spent a bunch of time making sure it would produce good stack traces during failures, and I added Async + request cancellation. It looks like you just integrated that a few hours ago.

If you would make a PR with any missing pieces of the LSP that you have added, that would be great.
